diff --git a/lib/socket.c b/lib/socket.c
index 63ca24f..11bd276 100644
--- a/lib/socket.c
+++ b/lib/socket.c
@@ -77,6 +77,7 @@
 #include "ucode/platform.h"
 
 #if defined(__linux__)
+#include <linux/version.h>
 # include <linux/if_packet.h>
 
 # ifndef SO_TIMESTAMP_OLD
@@ -1396,6 +1397,7 @@ static struct_t st_tpacket_auxdata = {
 	}
 };
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 0)
 static struct_t st_fanout_args = {
 	.size = sizeof(struct fanout_args),
 	.members = (member_t []){
@@ -1405,6 +1407,7 @@ static struct_t st_fanout_args = {
 		{ 0 }
 	}
 };
+#endif
 
 struct timeval_old_local {
 	long tv_sec;
@@ -1625,7 +1628,9 @@ static sockopt_t sockopts[] = {
 	{ SOL_PACKET, PACKET_ADD_MEMBERSHIP, &st_packet_mreq },
 	{ SOL_PACKET, PACKET_DROP_MEMBERSHIP, &st_packet_mreq },
 	{ SOL_PACKET, PACKET_AUXDATA, SV_BOOL },
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 0)
 	{ SOL_PACKET, PACKET_FANOUT, &st_fanout_args },
+#endif
 	{ SOL_PACKET, PACKET_LOSS, SV_BOOL },
 	{ SOL_PACKET, PACKET_RESERVE, SV_INT },
 	{ SOL_PACKET, PACKET_RX_RING, &st_tpacket_req },
@@ -5091,7 +5096,9 @@ void uc_module_init(uc_vm_t *vm, uc_value_t *scope)
 	ADD_CONST(PACKET_ADD_MEMBERSHIP);
 	ADD_CONST(PACKET_DROP_MEMBERSHIP);
 	ADD_CONST(PACKET_AUXDATA);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 0)
 	ADD_CONST(PACKET_FANOUT);
+#endif
 	ADD_CONST(PACKET_LOSS);
 	ADD_CONST(PACKET_RESERVE);
 	ADD_CONST(PACKET_RX_RING);
