diff --git a/candm.h b/candm.h
index dd99413..37ba2b9 100644
--- a/candm.h
+++ b/candm.h
@@ -575,6 +575,7 @@ typedef struct {
   Float orig_latest_meas;
   Float latest_meas;
   Float latest_meas_err;
+  uint16_t timeout_counter;
   int32_t EOR;
 } RPY_Source_Data;
 
diff --git a/client.c b/client.c
index ed551c6..620d7fd 100644
--- a/client.c
+++ b/client.c
@@ -1980,7 +1980,7 @@ process_cmd_sources(char *line)
     printf("||                                 |    |           \\\n");
   }
 
-  print_header("MS Name/IP address         Stratum Poll Reach LastRx Last sample               ");
+  print_header("MS Name/IP address         Stratum Poll Reach LastRx Last sample timeout              ");
 
   /*           "MS NNNNNNNNNNNNNNNNNNNNNNNNNNN  SS  PP   RRR  RRRR  SSSSSSS[SSSSSSS] +/- SSSSSS" */
 
@@ -2041,7 +2041,7 @@ process_cmd_sources(char *line)
         break;
     }
 
-    print_report("%c%c %-27s  %2d  %2d   %3o  %I  %+S[%+S] +/- %S\n",
+    print_report("%c%c %-27s  %2d  %2d   %3o  %I  %+S[%+S] +/- %S  %5d\n",
                  mode_ch, state_ch, name,
                  ntohs(reply.data.source_data.stratum),
                  (int16_t)ntohs(reply.data.source_data.poll),
@@ -2050,6 +2050,7 @@ process_cmd_sources(char *line)
                  UTI_FloatNetworkToHost(reply.data.source_data.latest_meas),
                  UTI_FloatNetworkToHost(reply.data.source_data.orig_latest_meas),
                  UTI_FloatNetworkToHost(reply.data.source_data.latest_meas_err),
+                 ntohs(reply.data.source_data.timeout_counter),
                  REPORT_END);
   }
 
diff --git a/cmdmon.c b/cmdmon.c
index 89f5b95..0a4fc48 100644
--- a/cmdmon.c
+++ b/cmdmon.c
@@ -631,6 +631,7 @@ handle_source_data(CMD_Request *rx_message, CMD_Reply *tx_message)
     tx_message->data.source_data.orig_latest_meas = UTI_FloatHostToNetwork(report.orig_latest_meas);
     tx_message->data.source_data.latest_meas = UTI_FloatHostToNetwork(report.latest_meas);
     tx_message->data.source_data.latest_meas_err = UTI_FloatHostToNetwork(report.latest_meas_err);
+    tx_message->data.source_data.timeout_counter = htons(report.timeout_counter);
   } else {
     tx_message->status = htons(STT_NOSUCHSOURCE);
   }
diff --git a/ntp_core.c b/ntp_core.c
index b73309e..a9ac2fd 100644
--- a/ntp_core.c
+++ b/ntp_core.c
@@ -959,6 +959,8 @@ receive_timeout(void *arg)
 
   DEBUG_LOG("Receive timeout for %s", UTI_IPSockAddrToString(&inst->remote_addr));
 
+  SRC_IncreaseTimeoutCounter(inst->source);
+
   inst->rx_timeout_id = 0;
   close_client_socket(inst);
 }
@@ -2055,6 +2057,8 @@ process_response(NCR_Instance inst, NTP_Local_Address *local_addr,
 
     SRC_UpdateReachability(inst->source, synced_packet);
 
+    SRC_ResetTimeoutCounter(inst->source);
+
     if (synced_packet) {
       if (inst->copy && inst->remote_stratum > 0) {
         /* Assume the reference ID and stratum of the server */
@@ -2165,7 +2169,7 @@ process_response(NCR_Instance inst, NTP_Local_Address *local_addr,
             NTP_LVM_TO_MODE(message->lvm), interleaved_packet ? 'I' : 'B',
             tss_chars[local_transmit.source],
             tss_chars[local_receive.source]);
-  }            
+  }
 
   return good_packet;
 }
diff --git a/reports.h b/reports.h
index c403c23..654e975 100644
--- a/reports.h
+++ b/reports.h
@@ -50,6 +50,7 @@ typedef struct {
   double orig_latest_meas; /* seconds */
   double latest_meas; /* seconds */
   double latest_meas_err; /* seconds */
+  uint16_t timeout_counter;
 } RPT_SourceReport ;
 
 typedef struct {
diff --git a/sources.c b/sources.c
index c439a74..0412196 100644
--- a/sources.c
+++ b/sources.c
@@ -139,6 +139,9 @@ struct SRC_Instance_Record {
 
   /* Flag indicating the source has a leap second vote */
   int leap_vote;
+
+  /* Source timeout counter */
+  uint16_t timeout_counter;
 };
 
 /* ================================================== */
@@ -321,6 +324,7 @@ SRC_ResetInstance(SRC_Instance instance)
   instance->stratum = 0;
   instance->leap = LEAP_Unsynchronised;
   instance->leap_vote = 0;
+  instance->timeout_counter = 0;
 
   memset(&instance->sel_info, 0, sizeof (instance->sel_info));
 
@@ -460,6 +464,21 @@ special_mode_end(void)
     return 1;
 }
 
+void
+SRC_ResetTimeoutCounter(SRC_Instance inst)
+{
+  inst->timeout_counter = 0;
+}
+
+void
+SRC_IncreaseTimeoutCounter(SRC_Instance inst)
+{
+  if (inst->timeout_counter < UINT16_MAX)
+    inst->timeout_counter++;
+  else
+    inst->timeout_counter = UINT16_MAX;
+}
+
 void
 SRC_UpdateReachability(SRC_Instance inst, int reachable)
 {
@@ -1604,6 +1623,7 @@ SRC_ReportSource(int index, RPT_SourceReport *report, struct timespec *now)
     }
 
     report->reachability = src->reachability;
+    report->timeout_counter = src->timeout_counter;
 
     /* Call stats module to fill out estimates */
     SST_DoSourceReport(src->stats, report, now);
diff --git a/sources.h b/sources.h
index 6c799bb..e955cda 100644
--- a/sources.h
+++ b/sources.h
@@ -102,6 +102,10 @@ extern void SRC_UnsetActive(SRC_Instance inst);
 /* This routine updates the reachability register */
 extern void SRC_UpdateReachability(SRC_Instance inst, int reachable);
 
+/* Those routines update the source timeout counter*/
+void SRC_ResetTimeoutCounter(SRC_Instance inst);
+void SRC_IncreaseTimeoutCounter(SRC_Instance inst);
+
 /* This routine marks the source unreachable */
 extern void SRC_ResetReachability(SRC_Instance inst);
