diff --git a/libfstools/common.c b/libfstools/common.c
index 07cf39d..41418c6 100644
--- a/libfstools/common.c
+++ b/libfstools/common.c
@@ -199,8 +199,10 @@ char *block_get_slave(const char *bdev)
 		goto out;
 
 	while ((d = xreaddir(dirp))) {
-		if (++slaves > 1)
-			goto out;	/* more than one slave */
+		if (++slaves > 1) {
+			ULOG_WARN("More than one slave => take the first one");
+			break; /* more than one slave */
+		}
 		strncpy(slave, d->d_name, sizeof(slave));
 		slave[sizeof(slave) - 1] = '\0'; /* make sure slave is always NULL terminated */
 	}
@@ -210,5 +212,5 @@ char *block_get_slave(const char *bdev)
 		closedir(dirp);
 	if  (fd >= 0)
 		close(fd);
-	return slaves == 1 ? slave : NULL;
+	return slaves > 0 ? slave : NULL;
 }
