diff --git a/src/tool_operate.c b/src/tool_operate.c
index f6ae64dd3..47c457759 100644
--- a/src/tool_operate.c
+++ b/src/tool_operate.c
@@ -1674,47 +1674,7 @@ static CURLcode single_transfer(struct GlobalConfig *global,
           my_setopt_str(curl, CURLOPT_CERTINFO, 1L);
 
         if(feature_ssl) {
-          /* Check if config->cert is a PKCS#11 URI and set the
-           * config->cert_type if necessary */
-          if(config->cert) {
-            if(!config->cert_type) {
-              if(is_pkcs11_uri(config->cert)) {
-                config->cert_type = strdup("ENG");
-              }
-            }
-          }
-
-          /* Check if config->key is a PKCS#11 URI and set the
-           * config->key_type if necessary */
-          if(config->key) {
-            if(!config->key_type) {
-              if(is_pkcs11_uri(config->key)) {
-                config->key_type = strdup("ENG");
-              }
-            }
-          }
-
-          /* Check if config->proxy_cert is a PKCS#11 URI and set the
-           * config->proxy_type if necessary */
-          if(config->proxy_cert) {
-            if(!config->proxy_cert_type) {
-              if(is_pkcs11_uri(config->proxy_cert)) {
-                config->proxy_cert_type = strdup("ENG");
-              }
-            }
-          }
-
-          /* Check if config->proxy_key is a PKCS#11 URI and set the
-           * config->proxy_key_type if necessary */
-          if(config->proxy_key) {
-            if(!config->proxy_key_type) {
-              if(is_pkcs11_uri(config->proxy_key)) {
-                config->proxy_key_type = strdup("ENG");
-              }
-            }
-          }
-
-          /* In debug build of curl tool, using
+         /* In debug build of curl tool, using
            *    --cert loadmem=<filename>:<password> --cert-type p12
            *  must do the same thing as classic:
            *    --cert <filename>:<password> --cert-type p12
