--- a/pppd/sys-linux.c
+++ b/pppd/sys-linux.c
@@ -172,7 +172,6 @@
 static int tty_disc = N_TTY;	/* The TTY discipline */
 static int ppp_disc = N_PPP;	/* The PPP discpline */
 static int initfdflags = -1;	/* Initial file descriptor flags for fd */
-static int ppp_fd = -1;		/* fd which is set to PPP discipline */
 static int sock_fd = -1;	/* socket for doing interface ioctls */
 static int slave_fd = -1;	/* pty for old-style demand mode, slave */
 static int master_fd = -1;	/* pty for old-style demand mode, master */
@@ -180,6 +179,8 @@
 static int sock6_fd = -1;
 #endif /* INET6 */
 
+int ppp_fd = -1;		/* fd which is set to PPP discipline */
+
 /*
  * For the old-style kernel driver, this is the same as ppp_fd.
  * For the new-style driver, it is the fd of an instance of /dev/ppp
--- a/pppd/pppd.h
+++ a/pppd/pppd.h
@@ -498,6 +498,10 @@
 
 extern struct userenv *userenv_list;
 
+extern int ppp_fd;
+extern int ppp_dev_fd;
+extern u_char inpacket_buf[];
+
 /*
  * Moved from main.c
  */
