--- a/pppd/pppd.h
+++ b/pppd/pppd.h
@@ -868,6 +868,8 @@
 //extern void (*compression_protocol_hook)(char *);
 extern void (*authentication_protocol_hook)(char *);
 extern void (*discovery_succeeded_hook)(void);
+extern void (*mru_changed_hook)(int);
+extern void (*mtu_changed_hook)(int);
 
 /* Let a plugin snoop sent and received packets.  Useful for L2TP */
 extern void (*snoop_recv_hook)(unsigned char *p, int len);
--- a/pppd/shared.c
+++ b/pppd/shared.c
@@ -65,6 +65,8 @@
 int (*new_phase_hook)(int) = NULL;
 void (*snoop_recv_hook)(unsigned char *p, int len) = NULL;
 void (*snoop_send_hook)(unsigned char *p, int len) = NULL;
+void (*mru_changed_hook)(int mru) = NULL;
+void (*mtu_changed_hook)(int mtu) = NULL;
 
 int conn_running;      /* we have a [dis]connector running */
 
@@ -714,6 +716,9 @@
 {
	int errs;
 
+	if (mtu_changed_hook)
+		mtu_changed_hook(mtu);
+
	if (the_channel->send_config == NULL)
		return 0;
	errs = error_count;
@@ -735,6 +742,9 @@
 {
	int errs;
 
+	if (mru_changed_hook)
+		mru_changed_hook(mru);
+
	if (the_channel->recv_config == NULL)
		return 0;
	errs = error_count;
