--- a/pppd/pppd.h
+++ b/pppd/pppd.h
@@ -58,7 +58,7 @@
 #include <sys/param.h>		/* for MAXPATHLEN and BSD4_4, if defined */
 #include <sys/types.h>		/* for u_int32_t, if defined */
 #include <sys/time.h>		/* for struct timeval */
+#include "ppp_defs.h"
-#include <net/ppp_defs.h>
 #include "patchlevel.h"

 #ifdef INET6
--- a/pppd/Makefile.linux
+++ b/pppd/Makefile.linux
@@ -24,7 +24,7 @@

 HEADERS = ccp.h session.h chap-new.h ecp.h fsm.h ipcp.h \
 	ipxcp.h lcp.h magic.h md5.h patchlevel.h pathnames.h pppd.h \
-	upap.h eap.h generator.h
+	upap.h eap.h generator.h ppp_defs.h

 MANPAGES = pppd.8
 PPPDOBJS = main.o magic.o fsm.o lcp.o ipcp.o upap.o chap-new.o md5.o ccp.o \
--- /dev/null
+++ b/pppd/ppp_defs.h
@@ -0,0 +1,11 @@
+#ifndef _NET_PPP_DEFS_H
+#define _NET_PPP_DEFS_H
+
+#define __need_time_t
+#include <time.h>
+
+#include <asm/types.h>
+#include <linux/ppp_defs.h>
+
+#define PPP_EAP		0xc227	/* Extensible Authentication Protocol */
+#endif

